pyogrio/_io.pyx

```diff
@@ -2178,3 +2180,182 @@ def ogr_write(
         exc = exc_check()
         if exc:
             raise DataSourceError(f"Failed to write features to dataset {path}; {exc}")
+
+
+def ogr_write_arrow(
+    str path,
+    str layer,
+    str driver,
+    arrow_obj,
+    str crs,
+    str geometry_type,
+    str geometry_name,
+    str encoding,
+    object dataset_kwargs,
+    object layer_kwargs,
+    bint append=False,
+    dataset_metadata=None,
+    layer_metadata=None,
+):
+    IF CTE_GDAL_VERSION < (3, 8, 0):
+        raise RuntimeError("Need GDAL>=3.8 for Arrow write support")
+
+    cdef OGRDataSourceH ogr_dataset = NULL
+    cdef OGRLayerH ogr_layer = NULL
+
+    if not encoding:
+        encoding = locale.getpreferredencoding()
+
+    layer_created = create_ogr_dataset_layer(
+        path, layer, driver, crs, geometry_type, encoding,
+        dataset_kwargs, layer_kwargs, append,
+        dataset_metadata, layer_metadata,
+        &ogr_dataset, &ogr_layer,
+    )
+    if geometry_name:
+        opts = {"GEOMETRY_NAME": geometry_name}
+    else:
+        opts = {}
+    cdef char **options = dict_to_options(opts)
+
+    stream_capsule = arrow_obj.__arrow_c_stream__()
```

Comment:
Ah, yes, we should of course check that the method is present to provide a better error message if not. Could also already do that at the Python level in `write_arrow`?

For validating that the method actually returns a capsule, I am not sure that is needed (it would be very strange for someone to use exactly that method name but not intent to implement the protocol). And in any case Python already will raise "ValueError: PyCapsule_GetPointer called with invalid PyCapsule object"

Reply:
I don't have a terribly strong opinion about where validation should occur, but our general tendency for the I/O functions is to do as much validation as possible (e.g., for parameters not being inspected / modified in the Python tier) in the lower-level Cython code rather than in the Python tier.